aiidalab_widgets_base/viewers.py

```diff
@@ -554,6 +555,59 @@ def _observe_all_representations(self, change):
         if change["new"]:
             self._all_representations[-1].viewer_class = self
 
+    def _compute_bonds(self, structure, radius=1.0, color="element", povray=False):
+        """Create an array of bonds for the structure."""
+
+        import ase.neighborlist
+
+        radius = radius / 5
+
+        cutoff = ase.neighborlist.natural_cutoffs(structure, mult=1.09)
+        bonds = []
+
+        def povray_cylinder(v1, v2, radius, atom, color):
+            color = np.array(Colors[atom.symbol])
+            return vapory.Cylinder(
+                v1,
+                v2,
+                radius,
+                vapory.Pigment("color", np.array(color)),
+                vapory.Finish("phong", 0.8, "reflection", 0.05),
+            )
+
+        def cylinder(v1, v2, radius, atom, color):
+            color = (
+                colors.jmol_colors[atom.number].tolist()
+                if color == "element"
+                else RGB_colors[color]
+            )
+            return (
+                "cylinder",
+                tuple(v1),
+                tuple(v2),
+                tuple(color),
+                radius,
+            )
+
+        # Choose the correct function for computing the cylinder.
+        compute_cylinder = povray_cylinder if povray else cylinder
+
+        if len(structure) > 1:
+            ii, jj, bond_lengths = ase.neighborlist.neighbor_list(
+                "ijD", structure, cutoff, self_interaction=False
+            )
+            for id1, id2, bond_length in zip(ii, jj, bond_lengths):
+                atom1, atom2 = structure[id1], structure[id2]
+
+                # Bond parameters
+                v1 = np.array([atom1.x, atom1.y, atom1.z])
+                half_bond_point = v1 + bond_length * Radius[atom1.symbol] / (
+                    Radius[atom1.symbol] + Radius[atom2.symbol]
+                )
```

Comment:
The math here is wrong. You need both the `vector` and the `length` to get the middle point.

Reply:
The naming here is confusing, I guess. This is to place the point where the bond changes colour. We find this point using the proportion of the atomic radius of the atoms.

I renamed the variable to `intermediate_point` in https://github.com/aiidalab/aiidalab-widgets-base/pull/535/commits/d4f0a063dc36c9c6409db73c2a6d326fa78e3fdc

Reply:
> We find this point using the proportion of the atomic radius of the atoms.

No. The correct math is:
```
# get the bond_vector from the `neighbor_list` too.
normal = np.linalg.norm(bond_vector)
r1 = Radius[atom1.symbol]
r2 = Radius[atom2.symbol]
intermediate_point = v1 + r1*normal + (bond_length - r1 - r2)/2*normal
# you can use the short term:  intermediate_point = v1 + (bond_length + r1 - r2)/2*normal
```

Reply:
I don't understand what is wrong with our approach.

`pos1` - location of the first atom
`pos2` - location of the second atom
`bond_vector` is a vector connecting the first atom and the second atom.
`r1` - is the vdW radius of the first atom
`r2` - is the vdW radius of the first atom

The intermediate point is a percentage of the vector r1 relative its contribution to the sum of two vdW radii.

`bond_vector = pos2 - pos1`

`atom1_contribution = r1 / (r1 + r2)`
`atom2_contribution = r2 / (r1 + r2)`

`intermediate_point = pos1 + bond_vector * atom1_contribution`

or

`intermediate_point = pos2 + bond_vector * atom2_contribution`

`atom1_contribution + atom2_contribution = r1 / (r1 + r2) + r2 / (r1 + r2) = (r1 + r2) / (r1 + f2)`

I'd appreciate it if you could take a moment to double-check our work before mentioning any concerns about the math. Thank you!

Reply:
For the record, we discussed with @superstar54 and @cpignedoli, where we agreed to cut the bond in the middle of the visible part.

Reply:
done in https://github.com/aiidalab/aiidalab-widgets-base/pull/535/commits/9ee0f991412f276cc77bb4890c717020e01fbe25